mastermind.py

```diff
@@ -0,0 +1,125 @@
+import random
+import string
+import matplotlib.pyplot as plt
+
+def generateRandomString(length):
+    characters = string.ascii_letters + string.digits + string.punctuation
+    randomString = ''.join(random.choice(characters) for _ in range(length))
+    return randomString
+
+def fitnessFunction(candidate, target):
+    return sum(1 for a, b in zip(candidate, target) if a == b)
+
+def selectParents(population, fitnessScores, numParents):
+    parents = sorted(zip(population, fitnessScores), key=lambda x: x[1], reverse=True)[:numParents]
+    return [p[0] for p in parents]
+
+def crossover(parent1, parent2):
+    crossoverPoint = random.randint(0, len(parent1) - 1)
+    return parent1[:crossoverPoint] + parent2[crossoverPoint:]
+
+def mutate(candidate, mutationRate):
+    characters = string.ascii_letters + string.digits + string.punctuation
+    candidateList = list(candidate)
+    for i in range(len(candidateList)):
+        if random.random() < mutationRate:
+            candidateList[i] = random.choice(characters)
+    return ''.join(candidateList)
```

Comment:
ez jo eros mutalas, hogy sok helyen is valtozhat, de vegulis... Lattal valami cikket, ahol igy csinaltak mutaciot amugy? Vagy sajat otlet?

Reply:
Valahol bele akartam csempészni a mutálást és úgy gondoltam, ha nem akarok random string darabokat újragenerálni, akkor karakterenként ez egy elfogadható ötlet. Tény, hogy nem túl hatékony, mivel minden karakterre minden crossoverben futtatásra kerül egy if és egy random generálás, ami futási időben exponenciálisan növekszik. A paraméterezhetősége miatt annyira nem mondanám erősnek, mivel, ha a mutationRate 0.001, akkor átlagosan csak minden ezredik karaktert módosítja.

Reply:
Az exponencialis dolgot nem latom, de ugyanaz mint a masik helyen: akkor ez sajat otlet, nem mashonnet vetted. Hogy jo-e, az meg ugye puding, de ha minden igaz, mukodott. (A targy szempontjabol amugy meg ez sem elvaras, hogy uberperformant legyen.)